File: diagonal.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
according to direction
store current element
if element is at the corner of array increase opposite of i or j and change direction
else if just reached edge and not corner increment opposite i or j and change dir
else decrement i or j and increment opposite i or j
do this for both directions
return
*/
package main

import "fmt"

func findDiagonalOrder(matrix [][]int) []int {
	m := len(matrix)
	if m == 0 {
		return []int{}
	}
	n := len(matrix[0])
	if n == 0 {
		return []int{}
	}
	res := make([]int, m*n)
	dir := 1
	i, j, ind := 0,0,0
	for ind < m*n {
		res[ind] = matrix[i][j]
		if dir == 1 {
			if j == n - 1 {
				i++
				dir = -1
			} else if i == 0 {
				j++
				dir = -1
			} else {
				i--
				j++
			}
		} else {
			if i == m - 1 {
				j++
				dir = 1
			} else if j == 0 {
				i++
				dir = 1
			} else {
				i++
				j--
			}
		}
		ind++
	}
	return res
}

func MainDiagonal() {
	fmt.Println(findDiagonalOrder([][]int{{ 1, 2, 3 },{ 4, 5, 6 }, { 7, 8, 9 }})) //expected:  [1,2,4,7,5,3,6,8,9]
}

File: src/Diagonal.java
// Time Complexity : O(MN)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english
/*
we Have two directions +1 and -1, which are just flags to monitor if we have to go
up or down. Then we are handling the edge cases and directions change in both as we traverse each diagonals and storing
the result in an array.
 */
// Your code here along with comments explaining your approach
public class Diagonal {
        public int[] findDiagonalOrder(int[][] matrix) {
            //base
            int m = matrix.length; int n = matrix[0].length;
            int result[] = new int[m*n];
            int k = 0;
            int dir=1;
            int i=0; int j=0;
            while(k < result.length) {
                result[k] = matrix[i][j];
                k++;
                //next i and j matrix
                if(dir == 1) {
                    if( j == n-1) {
                        i++; dir = -1;
                    } else if(i == 0) {
                        j++; dir = -1;
                    } else {
                        i--; j++;
                    }
                }
                else {
                    if( i == m-1) {
                        j++; dir = 1;
                    } else if(j == 0) {
                        i++; dir = 1;
                    } else {
                        j--; i++;
                    }
                }
            }
            return result;
        }
}
